src/Microsoft.AspNetCore.Blazor.Razor.Extensions/ComponentLoweringPass.cs

```diff
@@ -115,12 +124,166 @@ public override void VisitTagHelper(TagHelperIntermediateNode node)
 
             public override void VisitTagHelperBody(TagHelperBodyIntermediateNode node)
             {
+                // Wrap the component's children in a ChildContent node if we have some significant
+                // content.
+                if (node.Children.Count == 0)
+                {
+                    return;
+                }
+
+                // If we get a single HTML content node containing only whitespace,
+                // then this is probably a tag that looks like '<MyComponent>  </MyComponent>
+                //
+                // We don't want to create a child content for this case, because it can conflict
+                // with a child content that's set via an attribute. We don't want the formatting
+                // of insigificant whitespace to be annoying when setting attributes directly.
+                if (node.Children.Count == 1 && IsIgnorableWhitespace(node.Children[0]))
+                {
+                    return;
+                }
+
+                // From here we fork and behave differently based on whether the component's child content is
+                // implicit or explicit.
+                //
+                // Explict child content will look like: <MyComponent><ChildContent><div>...</div></ChildContent></MyComponent>
+                // compared with implicit: <MyComponent><div></div></MyComponent>
+                //
+                // Using implicit child content:
+                // 1. All content is grouped into a single child content lambda, and assiged to the property 'ChildContent'
+                //
+                // Using explicit child content:
+                // 1. All content must be contained within 'child content' elements that are direct children
+                // 2. Whitespace outside of 'child content' elements will be ignored (not an error)
+                // 3. Non-whitespace outside of 'child content' elements will cause an error
+                // 4. All 'child content' elements must match parameters on the component (exception for ChildContent,
+                //    which is always allowed.
```

Comment:
Rule 4 here is interesting.

I'm sure it's fine right now. If, in the future, we wanted to relax this so that we offered code completion for known child items, but also allowed unknown ones, how far would that diverge from the current implementation? This would allow us to make child items have the same semantics as normal attribute-style parameters, where you can also pass unknown things that show up like a name-value collection at runtime, and would allow things like:

```
<StatusCodeDisplay Status=currentStatus>
    <Http200>OK</Http200>
    <Http400>BadRequest</Http400>
    ... etc, without having to declare explicit [Parameter] properties for each possible item
</StatusCodeDisplay>
```

Not saying we need this right now. Just interested in whether we could extend to this in the future.

Reply:
As discussed on this mornings call, this would require component child content to operate in a bi-modal fashion. We'd allow component authors to opt-in to a model where child content must be explicit, but arbitrary elements are accepted. 

Anyone who needs a feature like this today can accomplish this using the runtime (make a `Dictionary<string, RenderFragment>` parameter.